garbage_collector.rst

```diff
@@ -0,0 +1,449 @@
+
+    >>> container = []
+    >>> container.append(container)
+    >>> sys.getrefcount(container)
+    3
+    >>> del container
+
+In this example, ``container`` holds a reference to itself, so even when we remove
+our reference to it (the variable "container") the reference count never falls to 0
+because it still has its own internal reference and therefore it will never be
+cleaned just by simple reference counting. For this reason some additional machinery
+is needed to clean these reference cycles between objects once they become
+unreachable. We normally refer to this additional machinery as the Garbage Collector,
+but technically reference counting is also a form of garbage collection.
+
+Memory layout and object structure
+----------------------------------
+
+Normally the C structure supporting a regular Python object looks as follows:
+
+.. code-block:: none
+
+    object -----> +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+ \
+                  |                    ob_refcnt                  | |
+                  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+ | PyObject_HEAD
+                  |                    *ob_type                   | |
+                  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+ /
+                  |                      ...                      |
+                  
+
+In order to support the garbage collector, the memory layout of objects is altered
+to accommodate extra information **before** the normal layout:
+
+.. code-block:: none
+
+                  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+ \
+                  |                    *_gc_next                  | |
+                  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+ | PyGC_Head
+                  |                    *_gc_prev                  | |
+    object -----> +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+ /
+                  |                    ob_refcnt                  | \
+                  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+ | PyObject_HEAD
+                  |                    *ob_type                   | |
+                  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+ /
+                  |                      ...                      |
+                  
+
+In this way the object can be treated as a normal python object and when the extra
+information associated to the GC is needed the previous fields can be accessed by a
+simple type cast from the original object: :code:`((PyGC_Head *)(the_object)-1)`.
+
+As is explained later in the `Optimization: reusing fields to save memory`_ section,
+these two extra fields are normally used to keep a double linked list of all the
+objects tracked by the garbage collector (these lists are the GC generations, more on
+that in the `Optimization: reusing fields to save memory`_ section) but they are also
+reused to fullfill other pourposes when the full double linked list structure is not
+needed as a memory optimization.
+
```

Comment:
```suggestion

Doubly linked lists are used because they efficiently support most frequently required operations.  In
general, the collection of all objects tracked by GC are partitioned into disjoint sets, each in its own
doubly linked list.  Between collections, objects are partitioned into "generations", reflecting how
often they're survived collection attempts.  During collections, the generations(s) being collected
are further partitioned into, e.g., sets of reachable and unreachable objects.  Doubly linked lists
support moving an object from one partition to another, adding a new object,  removing an object
entirely (objects tracked by GC are most often reclaimed by the refcounting system when GC
isn't running at all!), and merging partitions, all with a small constant number of pointer updates.